README.md
# norm

昆仑虚团队项目规范

校验 commit-msg 前也会先校验 eslint、stylelint
cz 也会事先校验 code

release 脚本自动生成 changelog tag, 并允许自动发布

## 初始化

1. 1 执行 `npx klx-norm` 并选择要配置内容

## 使用

### 1. eslint

```diff
// .eslintrc.js
+ const path = require("path");

module.exports = {
+ extends: [path.resolve(__dirname, "./node_modules/@kunlunxu/norm/.eslintrc.js")],
};
```

### 2. stylelintrc

```diff
// .stylelintrc.js
+ const path = require("path");

module.exports = {
+ extends: [path.resolve(__dirname, "./node_modules/@kunlunxu/norm/.stylelintrc.js")],
};
```

### 3. .vscode/setting.json 自动格式化配置

```diff
// .vscode/settings.json
+ {
+   // -----------统一配置--------------
+   "editor.codeActionsOnSave": { // 代码自动格式配置
+     "source.fixAll.eslint": true, // 根据 eslint 配置进行格式化, 需要安装 eslint 插件
+     "source.fixAll.stylelint": true // 根据 stylelint 配置进行格式化, 需要安装 stylelint 插件
+   },
+   // stylelint 插件允许检查的文件
+   "stylelint.validate": [
+     "css",
+     "less",
+     "postcss",
+     "scss",
+     "sass",
+   ],
+ }
```

## TODO

[] eslint 所有规则校验、配置过一遍、同时需要验证 vscode 自动修复是否可用
[] eslint jsx 属性格式、引入配置
[] prettier
[] 日志颜色问题


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2.8.1](https://github.com/KunLunXu0-0/norm/compare/v2.8.0...v2.8.1) (2022-08-10)

## [2.8.0](https://github.com/KunLunXu0-0/norm/compare/v2.7.2...v2.8.0) (2022-08-03)

### [2.7.2](https://github.com/KunLunXu0-0/norm/compare/v2.7.1...v2.7.2) (2022-08-03)

### [2.7.1](https://github.com/KunLunXu0-0/norm/compare/v2.7.0...v2.7.1) (2022-08-02)

## [2.7.0](https://github.com/KunLunXu0-0/norm/compare/v2.6.1...v2.7.0) (2022-08-02)


### Features

* 🎸 update eslint-react hooks ([e10c7da](https://github.com/KunLunXu0-0/norm/commit/e10c7da6fdee2851b2cdc549535d94ee1b5d3a04))

### [2.6.1](https://github.com/KunLunXu0-0/norm/compare/v2.6.0...v2.6.1) (2022-07-31)

## [2.6.0](https://github.com/KunLunXu0-0/norm/compare/v2.5.8...v2.6.0) (2022-07-30)


### Features

* 🎸 add ts ([82c45e3](https://github.com/KunLunXu0-0/norm/commit/82c45e3397b7f453ab1b8de1cb9166accf24f974))

### [2.5.8](https://github.com/KunLunXu0-0/norm/compare/v2.5.7...v2.5.8) (2022-07-18)

### [2.5.7](https://github.com/KunLunXu0-0/norm/compare/v2.5.6...v2.5.7) (2022-07-18)


### Features

* 🎸 update bin/norm ([c8e13db](https://github.com/KunLunXu0-0/norm/commit/c8e13db395eaa326255fdb91f1355034a665be02))

### [2.5.6](https://github.com/KunLunXu0-0/norm/compare/v2.5.5...v2.5.6) (2022-07-16)


### Features

* 🎸 update npm ([71f0d4c](https://github.com/KunLunXu0-0/norm/commit/71f0d4c5d577d435df4d89ae45fa89c11c668b4b))

### [2.5.5](https://github.com/KunLunXu0-0/norm/compare/v2.5.4...v2.5.5) (2022-07-10)

### [2.5.4](https://github.com/KunLunXu0-0/norm/compare/v2.5.3...v2.5.4) (2022-07-10)

### [2.5.3](https://github.com/KunLunXu0-0/norm/compare/v2.5.2...v2.5.3) (2022-07-10)

### [2.5.2](https://github.com/KunLunXu0-0/norm/compare/v2.5.1...v2.5.2) (2022-07-10)

### [2.5.1](https://github.com/KunLunXu0-0/norm/compare/v2.5.0...v2.5.1) (2022-07-09)

## [2.5.0](https://github.com/KunLunXu0-0/norm/compare/v2.4.0...v2.5.0) (2022-07-09)

## [2.4.0](https://github.com/qianyin925/qy-norm/compare/v2.3.0...v2.4.0) (2022-01-16)

## [2.3.0](https://github.com/qianyin925/qy-norm/compare/v2.2.0...v2.3.0) (2022-01-16)

## [2.2.0](https://github.com/qianyin925/qy-norm/compare/v2.1.0...v2.2.0) (2022-01-16)

## [2.1.0](https://github.com/qianyin925/qy-norm/compare/v2.0.0...v2.1.0) (2022-01-16)


### Features

* 🎸 【【qy-norm】Eslint index 针对 iswitch 进行调整】 https://www.tapd.cn/58837035/prong/stories/view/1158837035001000745 ([1234506](https://github.com/qianyin925/qy-norm/commit/1234506270408d5ec5408ef3ffd5b092b61f5a74))
* 🎸 【【qy-norm】Eslint index 针对 iswitch 进行调整】 https://www.tapd.cn/58837035/prong/stories/view/1158837035001000745 ([7afd77e](https://github.com/qianyin925/qy-norm/commit/7afd77e8eeada06530197971d16f7fc273f8e3c5))

## [2.0.0](https://github.com/qianyin925/qy-norm/compare/v1.5.0...v2.0.0) (2022-01-16)

## [1.5.0](https://github.com/qianyin925/qy-norm/compare/v1.4.9...v1.5.0) (2022-01-16)


### Features

* 【【all】commit 规范添加表情】 https://www.tapd.cn/58837035/prong/stories/view/1158837035001000682 ([36f166e](https://github.com/qianyin925/qy-norm/commit/36f166ecc24d32b3f08b77647e68601408ae328c))

### [1.4.9](https://github.com/qianyin925/qy-norm/compare/v1.4.8...v1.4.9) (2021-09-19)

### [1.4.8](https://github.com/qianyin925/qy-norm/compare/v1.4.7...v1.4.8) (2021-09-19)

### [1.4.7](https://github.com/qianyin925/qy-norm/compare/v1.4.6...v1.4.7) (2021-09-19)


### Features

* 🎸 update stylelint ([729bf16](https://github.com/qianyin925/qy-norm/commit/729bf16af46361641fdd64586864b372f22a1af8))

### [1.4.6](https://github.com/qianyin925/qy-norm/compare/v1.4.0...v1.4.6) (2021-09-19)


### Features

* 🎸 update cz ([c242844](https://github.com/qianyin925/qy-norm/commit/c242844e6c99553374f8762c4f1a00239c7350a2))


### Bug Fixes

* 🐛 同时执行多个初始脚本, 有的命令会失败 ([083b807](https://github.com/qianyin925/qy-norm/commit/083b8079acde4b5510037451a345a9c4e1fd7666))
* 🐛 qy-norm 执行总是失败 ([55ece20](https://github.com/qianyin925/qy-norm/commit/55ece2062919e3011ccd4aff332a18c02babb719))

## [1.4.0](https://github.com/qianyin925/qy-norm/compare/v1.3.0...v1.4.0) (2021-08-13)

## [1.3.0](https://github.com/qianyin925/qy-norm/compare/v1.2.0...v1.3.0) (2021-08-13)

## [1.2.0](https://github.com/qianyin925/qy-norm/compare/v1.1.0...v1.2.0) (2021-08-13)

## [1.1.0](https://github.com/qianyin925/qy-norm/compare/v1.0.15...v1.1.0) (2021-08-13)

### [1.0.15](https://github.com/qianyin925/qy-norm/compare/v1.0.14...v1.0.15) (2021-08-13)

### [1.0.14](https://github.com/qianyin925/qy-norm/compare/v1.0.13...v1.0.14) (2021-08-13)

### [1.0.13](https://github.com/qianyin925/qy-norm/compare/v1.0.12...v1.0.13) (2021-08-13)

### [1.0.12](https://github.com/qianyin925/qy-norm/compare/v1.0.11...v1.0.12) (2021-07-29)

### [1.0.11](https://github.com/qianyin925/qy-norm/compare/v1.0.10...v1.0.11) (2021-07-29)

### [1.0.10](https://github.com/qianyin925/qy-norm/compare/v1.0.9...v1.0.10) (2021-07-29)

### [1.0.9](https://github.com/qianyin925/qy-norm/compare/v1.0.8...v1.0.9) (2021-07-29)


### Bug Fixes

* 🐛 npm run cz has fix ([acab3dc](https://github.com/qianyin925/qy-norm/commit/acab3dc801913366143e1f21f1cd61626ed3487a))

### [1.0.8](https://github.com/qianyin925/qy-norm/compare/v1.0.7...v1.0.8) (2021-07-29)


### Bug Fixes

* 🐛 qy-commitlint 引入错误 ([d626bb0](https://github.com/qianyin925/qy-norm/commit/d626bb0157cfa2337d291305e8103a75e6703ea1))

### [1.0.7](https://github.com/qianyin925/qy-norm/compare/v1.0.6...v1.0.7) (2021-07-29)


### Features

* 🎸 remove  example ([ebb714a](https://github.com/qianyin925/qy-norm/commit/ebb714a27abae051eab8e4dd0bf953c6297c4641))

### [1.0.6](https://github.com/qianyin925/qy-norm/compare/v1.0.5...v1.0.6) (2021-07-29)


### Features

* 🎸 update script ([6e692a1](https://github.com/qianyin925/qy-norm/commit/6e692a1023f44719a97c2ad252a2b25c32b661a5))


.eslintrc.ts.js
module.exports = {
  // root: true,                  // 作用的目录是根目录
  parser: '@typescript-eslint/parser',
  parserOptions: {
    ecmaVersion: 'latest',
    sourceType: 'module',      // 按照模块的方式解析
    ecmaFeatures: {            // 想使用的额外的语言特性
      jsx: true,
      experimentalObjectRestSpread: true,
    },
  },
  env: {
    es6: true,                // es6 环境
    node: true,               // node 环境
    browser: true,            // 开发环境配置表示可以使用浏览器的方法
  },
  globals: {                 // 自定义全局变量
    _: true,
    _DEV_: true,
    lodash: true,
  },

  // 插件只是为 eslint 提供相关的规则配置, 要想使用还需要手动在 rules 中进行配置, 或者在 extends 中继承插件中的默认配置
  plugins: [
    'jsdoc',
    'react',
    'import',
    'react-hooks',
    '@typescript-eslint',
  ],
  extends: [
    'eslint:recommended',
    'plugin:react/jsx-runtime',
    'plugin:jsdoc/recommended',
    'plugin:import/typescript',
    'plugin:react-hooks/recommended',
    'plugin:@typescript-eslint/recommended',
  ],
  settings: {
    react: { version: 'detect' },
  },

  rules: {
    // 1. react 插件规则个性化设置: https://www.npmjs.com/package/eslint-plugin-react
    'react/default-props-match-prop-types': 2, // 对于必须 props 无需设置默认值
    'react/prop-types': 0,                     // 不强制要求设置 props 校验
    'react/display-name': 0,                   // 不强制设置 displayName(为组件命名)
    'react/self-closing-comp': 2,              // 不允许额外的结束标签, 如果没有子元素应该使用自闭合标签
    'react/jsx-closing-bracket-location': [    // 限制 JSX 中的右括号位置
      2,
      {
        nonEmpty: 'after-props',                // 非空则结束符跟在 props 后面
        selfClosing: 'line-aligned',            // 自动关闭, 则一行显示
      },
    ],
    'react/jsx-first-prop-new-line': 2,                  // 如果 JSX 标签占用多行并且有多个属性，则第一个属性应始终放在新行
    'react/jsx-max-props-per-line': [1, { maximum: 1 }], // 限制一行最多只能有一个 props

    'react/jsx-no-useless-fragment': 2,           // 禁止不必要的空标签 <></> || <React.Fragment>
    'react/jsx-indent': [2, 2],                   // JSX 缩进
    'react/jsx-equals-spacing': [2, 'never'],     // 在 JSX 属性 = 周围不能有空格
    'react/jsx-curly-spacing': 2,                 // 属性值 {} 和值不允许有空格 xxx={name}
    'react/jsx-one-expression-per-line': 2,       // JSX 中每行限制为一个表达式
    'react/jsx-props-no-multi-spaces': 2,         // 多个 props 之间不允许有多余的空格、行
    'react/jsx-tag-spacing': 2,                   // 验证 JSX < 和 > 周围的空格
    'react/jsx-wrap-multilines': [                // 多行 jsx 需要使用 () 包裹
      2,
      {
        declaration: 'parens-new-line',
        assignment: 'parens-new-line',
        return: 'parens-new-line',
        arrow: 'parens-new-line',
        condition: 'parens-new-line',
        logical: 'parens-new-line',
        prop: 'parens-new-line',
      },
    ],


    // 2. 常规个性化定制
    'indent': [2, 2, { SwitchCase: 1 }], // 缩进: 2个空格
    'max-len': [                         // 最大长度
      2,
      {
        code: 80,                        // 单行不超过 80
        ignoreComments: true,            // 忽略注释
        ignoreUrls: true,                // 忽略 URL
        ignoreStrings: true,             // 忽略字符串
        ignoreTemplateLiterals: true,    // 忽略模板字符串
        ignoreRegExpLiterals: true,      // 忽略正则
      },
    ],
    'keyword-spacing': [2],                                      // 强制关键字周围空格的一致性
    'rest-spread-spacing': [2, 'never'],                         // 扩展运算符及其表达式之间要求有空格
    'comma-spacing': [2, { before: false, after: true }],        // 强制在逗号周围使用空格
    'switch-colon-spacing': [2, { after: true, before: false }], // 强制在 switch 的冒号左右有空格

    'semi': [2, 'always'],                      // 要求使用分号
    'semi-style': [2, 'last'],                  // 强制分号处于句子末尾
    'no-extra-semi': 2,                         // 禁止不必要的分号(使用多个分号)

    'quote-props': [2, 'consistent-as-needed'],   // 如果有属性名称要求使用引号，则所有的属性名称都要使用引号；否则，禁止所有的属性名称使用引号
    'no-extra-boolean-cast': 0,                   // 禁止不必要的布尔转换
    'no-var': 2,                                  // 禁止使用 var
    'space-infix-ops': 2,                         // 要求中缀操作符周围有空格
    'eol-last': 2,                                // 文件结尾空一行
    'no-multiple-empty-lines': 2,                 // 避免出现多个空行。 在文件末尾只允许空一行
    'padded-blocks': [2, 'never'],                // 不要故意留一些没必要的空白行
    'object-curly-spacing': [2, 'always'],        // 花括号 {} 里加空格
    'array-bracket-spacing': 2,                   // 方括号[]里不要加空格
    'space-in-parens': [2, 'never'],              // 圆括号里不要加空格
    'key-spacing': 2,                             // 在对象的属性中,  键值之间要有空格
    'no-trailing-spaces': 2,                      // 行末不要空格

    'no-new-object': 2,                           // 禁止使用 Object 构造函数: new Object()
    'object-shorthand': 2,                        // 要求对象字面量简写语法, {fun(){}}, 该规则不标记对象字面量中的箭头函数

    'no-array-constructor': 2,                    // 禁止使用 Array 构造函数
    'array-callback-return': 2,                   // 强制数组方法的回调函数中有 return 语句
    'prefer-destructuring': [2, { object: true, array: true }], // 优先使用数组和对象解构
    'quotes': [2, 'single', { avoidEscape: true }],             // 字符串使用要求使用单引号
    'prefer-template': 2,                                       // 建议使用模板字面量而非字符串连接
    'template-curly-spacing': [2, 'never'],                     // 禁止模板字符串中的嵌入表达式周围空格的使用
    'no-eval': 2,                                               // 禁用 eval()
    'no-useless-escape': 2,                                     // 禁用不必要的转义
    'func-style': [2, 'expression'],                            // 要求使用函数表达式而不是函数声明
    'no-loop-func': 2,                                          // 禁止循环中存在函数
    'prefer-rest-params': 2,                                    // 建议使用剩余参数代替 arguments
    'no-new-func': 2,                                           // 禁用Function构造函数

    'space-before-blocks': [2, 'always'],         // 要求语句块之前的空格
    'space-before-function-paren': [2, 'always'], // 要求函数圆括号之前有一个空格
    'no-param-reassign': [0],                     // 允许对函数参数再赋值、对参数对象修改属性(理论上应该是禁止的, 但是总有特殊情况, 后台比较多)
    'prefer-spread': 2,                           // 建议使用扩展语法而非.apply()
    'prefer-arrow-callback': 2,                   // 要求使用箭头函数作为回调
    'arrow-spacing': [2, { before: true, after: true }], // 要求箭头函数的箭头之前或之后有空格
    'arrow-parens': [2, 'always'],                       // 箭头函数中: 在可以省略括号的地方强制不使用括号
    'arrow-body-style': [2, 'as-needed'],                // 箭头函数中: 当大括号是可以省略的, 强制不使用它们
    'no-confusing-arrow': 2,                             // 禁止在可能与比较操作符相混淆的地方使用箭头函数
    'implicit-arrow-linebreak': [2, 'beside'],           // 禁止在箭头函数体之前出现换行

    // 箭头函数完
    'no-useless-constructor': 2,        // 禁用不必要的构造函数
    'no-dupe-class-members': 2,         // 不允许类成员中有重复的名称
    'no-duplicate-imports': 2,          // 禁止重复导入

    'spaced-comment': [2, 'always'],    // 注释前加一个空格
    'no-whitespace-before-property': 2, // 禁止属性前有空白
    'newline-per-chained-call': [2, { ignoreChainWithDepth: 2 }], // 要求方法链中每个调用都有一个换行符
    'comma-style': [2, 'last'],                                   // 不要前置逗号
    'comma-dangle': [2, 'always-multiline'],                      // 当最后一个元素或属性与闭括号 ] 或 } 在 不同的行时, 要求使用拖尾逗号；当在 同一行时, 禁止使用拖尾逗号。
    'no-new-wrappers': 2,                                         // 禁止对 String, Number 和 Boolean 使用 new 操作符
    'radix': [2, 'always'],                                       // 当使用parseInt()函数时要求使用第二参数
    'camelcase': [2],                                             // 强制使用驼峰
    'new-cap': 2,                                                 // 要求构造函数首字母大写

    'eqeqeq': 2,                                                  // 要求使用 === 1==
    'no-case-declarations': 2,                         // 禁止在 case 或 default 子句中出现词法声明
    'no-nested-ternary': 2,                            // 禁止使用嵌套的三元表达式
    'no-unneeded-ternary': 2,                          // 禁止可以表达为更简单结构的三元操作符
    'no-mixed-operators': 2,                           // 禁止混合使用不同的操作符, 需要加()
    'nonblock-statement-body-position': [2, 'beside'], // 强制单个语句禁止出现换行需要的话要加 {}
    'brace-style': 2,                                  // 强制在代码块中使用一致的大括号风格
    'no-else-return': 2,                               // 禁止在 else 前有 return

    'no-undef': 2,                                     // 禁用未声明的变量
    'one-var': [2, 'never'],                           // 要求每个作用域有多个变量声明时每个变量都需要独立的 const let 进行声明
    'prefer-const': 2,                                 // 建议使用const： 如果一个变量不会被重新赋值, 最好使用const进行声明
    'no-multi-assign': 2,                              // 禁止连续赋值(a=b=c=1)
    'no-plusplus': 2,                                  // 禁止使用 ++ --
    'no-unused-vars': 2,                               // 禁止出现未使用过的变量
    'no-undef-init': 2,                                // 不允许初始化变量值为 undefined
    'no-undefined': 2,                                 // 禁止将 undefined 作为标识符
    'dot-notation': 2,                                 // 要求尽可能使用点号, obj.xxx
    'padding-line-between-statements': [2,
      { blankLine: 'always', prev: 'block-like', next: '*' }, // 块语句之前需要空行
      { blankLine: 'always', prev: '*', next: 'block-like' }, // 块语句之后需要空行
      { blankLine: 'always', prev: 'export', next: '*' },     // export 语句之前需要空行
      { blankLine: 'always', prev: '*', next: 'export' },     // export 语句之后需要空行
    ],

    // 3. import
    'import/prefer-default-export': 2,  // 在只有一个导出的模块里, 用 export default
    'import/first': 2,                  // 10.7 import 放在其他所有语句之前
  },
};


.stylelintrc.js
// see: https://github.com/stylelint/stylelint/blob/master/docs/user-guide/rules/list.md
module.exports = {
  extends: ['stylelint-config-standard', 'stylelint-config-sass-guidelines', 'stylelint-config-standard-scss'],
  rules: {
    'max-nesting-depth': 3,  // 最大深度
    'selector-max-compound-selectors': 4, // 选择器最大个数 div .bar[data-val] > a.baz + .boom > #lorem {}
    'order/properties-alphabetical-order': null, // 关闭属性排序
    "declaration-empty-line-before": null,  // 属性间禁止空行
    'selector-pseudo-class-no-unknown': [true, {
      ignorePseudoClasses: ['global', 'local'],  // 允许使用 :global :local
    }],
    'selector-pseudo-element-no-unknown': [true, {
      ignorePseudoElements: ['global', 'local'], // 允许使用 :global :local
    }],
    // 选择器不允许限定类型
    'selector-no-qualifying-type': [true, {
      ignore: ["attribute", "class", "id"]
    }],
  },
  // 忽略文件
  ignoreFiles: ['**/*.js', '**/*.ts', '**/*.jsx', '**/*.tsx']
}


commitlint.config.js
// commit 校验
const types = [
  {
    desc: '其他类型(无法定位 type)、或者上一个 commit 补充代码',
    emoji: '🤖',
    value: 'chore',
  },
  {
    desc: '新增需求',
    emoji: '🎸',
    value: 'feat',
  },
  {
    desc: 'BUG 修复',
    emoji: '🐛',
    value: 'fix',
  },
  {
    desc: '重大功能重构',
    emoji: '💡',
    value: 'refactor',
  },
  {
    desc: '代码格式变更(不是 CSS 样式)',
    emoji: '💄',
    value: 'style',
  },
  {
    desc: '提高代码性能的变更',
    emoji: '⚡️',
    value: 'perf',
  },
  {
    desc: '持续集成和部署相关的改动(脚本、打包配置文件……)',
    emoji: '🎡',
    value: 'ci',
  },
  {
    desc: '项目文档变更',
    emoji: '📝',
    value: 'docs',
  },
  {
    desc: '新增(更新)测试用例',
    emoji: '💍',
    value: 'test',
  },
];

module.exports = {
  types,
  extends: ['@commitlint/config-conventional'],
  rules: {
    'type-enum': [2, 'always', types.map((v) => v.value)],
    'header-max-length': [0, 'always', 72],
    'body-max-length': [0, 'always', 72],
  },
};


.eslintrc.js
module.exports = {
  // root: true,                  // 作用的目录是根目录
  parserOptions: {
    ecmaVersion: 'latest',
    sourceType: 'module',      // 按照模块的方式解析
    ecmaFeatures: {            // 想使用的额外的语言特性
      jsx: true,
    },
  },
  env: {
    es6: true,                // es6 环境
    node: true,               // node 环境
    browser: true,            // 开发环境配置表示可以使用浏览器的方法
  },
  globals: {                 // 自定义全局变量
    _: true,
    _DEV_: true,
    lodash: true,
  },

  // 插件只是为 eslint 提供相关的规则配置, 要想使用还需要手动在 rules 中进行配置, 或者在 extends 中继承插件中的默认配置
  plugins: [
    'jsdoc',
    'react',
    'import',
    'react-hooks',
  ],
  extends: [
    'eslint:recommended',
    'plugin:react/jsx-runtime',
    'plugin:jsdoc/recommended',
    'plugin:import/recommended',
    'plugin:react-hooks/recommended',
  ],
  settings: {
    'react': { version: 'detect' },
    // 导入 import 配置
    'import/resolver': {
      node: {
        extensions: ['.js', '.jsx', '.ts', '.tsx', '.cjs', '.mjs'],
      },
    },
  },

  overrides: [
    // 针对 ts 进行处理
    {
      files: ['*.ts', '*.tsx'],
      plugins: ['@typescript-eslint'],
      parser: '@typescript-eslint/parser',
      extends: [
        'plugin:import/typescript',
        'plugin:@typescript-eslint/recommended',
      ],
    },
  ],

  rules: {
    // 1. react 插件规则个性化设置: https://www.npmjs.com/package/eslint-plugin-react
    'react/default-props-match-prop-types': 2, // 对于必须 props 无需设置默认值
    'react/prop-types': 0,                     // 不强制要求设置 props 校验
    'react/display-name': 0,                   // 不强制设置 displayName(为组件命名)
    'react/self-closing-comp': 2,              // 不允许额外的结束标签, 如果没有子元素应该使用自闭合标签
    'react/jsx-closing-bracket-location': [    // 限制 JSX 中的右括号位置
      2,
      {
        nonEmpty: 'after-props',                // 非空则结束符跟在 props 后面
        selfClosing: 'line-aligned',            // 自动关闭, 则一行显示
      },
    ],
    'react/jsx-first-prop-new-line': 2,                  // 如果 JSX 标签占用多行并且有多个属性，则第一个属性应始终放在新行
    'react/jsx-max-props-per-line': [1, { maximum: 1 }], // 限制一行最多只能有一个 props

    'react/jsx-no-useless-fragment': 2,           // 禁止不必要的空标签 <></> || <React.Fragment>
    'react/jsx-indent': [2, 2],                   // JSX 缩进
    'react/jsx-equals-spacing': [2, 'never'],     // 在 JSX 属性 = 周围不能有空格
    'react/jsx-curly-spacing': 2,                 // 属性值 {} 和值不允许有空格 xxx={name}
    'react/jsx-one-expression-per-line': 2,       // JSX 中每行限制为一个表达式
    'react/jsx-props-no-multi-spaces': 2,         // 多个 props 之间不允许有多余的空格、行
    'react/jsx-tag-spacing': 2,                   // 验证 JSX < 和 > 周围的空格
    'react/jsx-wrap-multilines': [                // 多行 jsx 需要使用 () 包裹
      2,
      {
        declaration: 'parens-new-line',
        assignment: 'parens-new-line',
        return: 'parens-new-line',
        arrow: 'parens-new-line',
        condition: 'parens-new-line',
        logical: 'parens-new-line',
        prop: 'parens-new-line',
      },
    ],


    // 2. 常规个性化定制
    'indent': [2, 2, { SwitchCase: 1 }], // 缩进: 2个空格
    'max-len': [                         // 最大长度
      2,
      {
        code: 80,                        // 单行不超过 80
        ignoreComments: true,            // 忽略注释
        ignoreUrls: true,                // 忽略 URL
        ignoreStrings: true,             // 忽略字符串
        ignoreTemplateLiterals: true,    // 忽略模板字符串
        ignoreRegExpLiterals: true,      // 忽略正则
      },
    ],
    'keyword-spacing': [2],                                      // 强制关键字周围空格的一致性
    'rest-spread-spacing': [2, 'never'],                         // 扩展运算符及其表达式之间要求有空格
    'comma-spacing': [2, { before: false, after: true }],        // 强制在逗号周围使用空格
    'switch-colon-spacing': [2, { after: true, before: false }], // 强制在 switch 的冒号左右有空格

    'semi': [2, 'always'],                      // 要求使用分号
    'semi-style': [2, 'last'],                  // 强制分号处于句子末尾
    'no-extra-semi': 2,                         // 禁止不必要的分号(使用多个分号)

    'quote-props': [2, 'consistent-as-needed'],   // 如果有属性名称要求使用引号，则所有的属性名称都要使用引号；否则，禁止所有的属性名称使用引号
    'no-extra-boolean-cast': 0,                   // 禁止不必要的布尔转换
    'no-var': 2,                                  // 禁止使用 var
    'space-infix-ops': 2,                         // 要求中缀操作符周围有空格
    'eol-last': 2,                                // 文件结尾空一行
    'no-multiple-empty-lines': 2,                 // 避免出现多个空行。 在文件末尾只允许空一行
    'padded-blocks': [2, 'never'],                // 不要故意留一些没必要的空白行
    'object-curly-spacing': [2, 'always'],        // 花括号 {} 里加空格
    'array-bracket-spacing': 2,                   // 方括号[]里不要加空格
    'space-in-parens': [2, 'never'],              // 圆括号里不要加空格
    'key-spacing': 2,                             // 在对象的属性中,  键值之间要有空格
    'no-trailing-spaces': 2,                      // 行末不要空格

    'no-new-object': 2,                           // 禁止使用 Object 构造函数: new Object()
    'object-shorthand': 2,                        // 要求对象字面量简写语法, {fun(){}}, 该规则不标记对象字面量中的箭头函数

    'no-array-constructor': 2,                    // 禁止使用 Array 构造函数
    'array-callback-return': 2,                   // 强制数组方法的回调函数中有 return 语句
    'prefer-destructuring': [2, { object: true, array: true }], // 优先使用数组和对象解构
    'quotes': [2, 'single', { avoidEscape: true }],             // 字符串使用要求使用单引号
    'prefer-template': 2,                                       // 建议使用模板字面量而非字符串连接
    'template-curly-spacing': [2, 'never'],                     // 禁止模板字符串中的嵌入表达式周围空格的使用
    'no-eval': 2,                                               // 禁用 eval()
    'no-useless-escape': 2,                                     // 禁用不必要的转义
    'func-style': [2, 'expression'],                            // 要求使用函数表达式而不是函数声明
    'no-loop-func': 2,                                          // 禁止循环中存在函数
    'prefer-rest-params': 2,                                    // 建议使用剩余参数代替 arguments
    'no-new-func': 2,                                           // 禁用Function构造函数

    'space-before-blocks': [2, 'always'],         // 要求语句块之前的空格
    'space-before-function-paren': [2, 'always'], // 要求函数圆括号之前有一个空格
    'no-param-reassign': [0],                     // 允许对函数参数再赋值、对参数对象修改属性(理论上应该是禁止的, 但是总有特殊情况, 后台比较多)
    'prefer-spread': 2,                           // 建议使用扩展语法而非.apply()
    'prefer-arrow-callback': 2,                   // 要求使用箭头函数作为回调
    'arrow-spacing': [2, { before: true, after: true }], // 要求箭头函数的箭头之前或之后有空格
    'arrow-parens': [2, 'always'],                       // 箭头函数中: 在可以省略括号的地方强制不使用括号
    'arrow-body-style': [2, 'as-needed'],                // 箭头函数中: 当大括号是可以省略的, 强制不使用它们
    'no-confusing-arrow': 2,                             // 禁止在可能与比较操作符相混淆的地方使用箭头函数
    'implicit-arrow-linebreak': [2, 'beside'],           // 禁止在箭头函数体之前出现换行

    // 箭头函数完
    'no-useless-constructor': 2,        // 禁用不必要的构造函数
    'no-dupe-class-members': 2,         // 不允许类成员中有重复的名称
    'no-duplicate-imports': 2,          // 禁止重复导入

    'spaced-comment': [2, 'always'],    // 注释前加一个空格
    'no-whitespace-before-property': 2, // 禁止属性前有空白
    'newline-per-chained-call': [2, { ignoreChainWithDepth: 2 }], // 要求方法链中每个调用都有一个换行符
    'comma-style': [2, 'last'],                                   // 不要前置逗号
    'comma-dangle': [2, 'always-multiline'],                      // 当最后一个元素或属性与闭括号 ] 或 } 在 不同的行时, 要求使用拖尾逗号；当在 同一行时, 禁止使用拖尾逗号。
    'no-new-wrappers': 2,                                         // 禁止对 String, Number 和 Boolean 使用 new 操作符
    'radix': [2, 'always'],                                       // 当使用parseInt()函数时要求使用第二参数
    'camelcase': [2],                                             // 强制使用驼峰
    'new-cap': 2,                                                 // 要求构造函数首字母大写

    'eqeqeq': 2,                                                  // 要求使用 === 1==
    'no-case-declarations': 2,                         // 禁止在 case 或 default 子句中出现词法声明
    'no-nested-ternary': 2,                            // 禁止使用嵌套的三元表达式
    'no-unneeded-ternary': 2,                          // 禁止可以表达为更简单结构的三元操作符
    'no-mixed-operators': 2,                           // 禁止混合使用不同的操作符, 需要加()
    'nonblock-statement-body-position': [2, 'beside'], // 强制单个语句禁止出现换行需要的话要加 {}
    'brace-style': 2,                                  // 强制在代码块中使用一致的大括号风格
    'no-else-return': 2,                               // 禁止在 else 前有 return

    'no-undef': 2,                                     // 禁用未声明的变量
    'one-var': [2, 'never'],                           // 要求每个作用域有多个变量声明时每个变量都需要独立的 const let 进行声明
    'prefer-const': 2,                                 // 建议使用const： 如果一个变量不会被重新赋值, 最好使用const进行声明
    'no-multi-assign': 2,                              // 禁止连续赋值(a=b=c=1)
    'no-plusplus': 2,                                  // 禁止使用 ++ --
    'no-unused-vars': 2,                               // 禁止出现未使用过的变量
    'no-undef-init': 2,                                // 不允许初始化变量值为 undefined
    'no-undefined': 2,                                 // 禁止将 undefined 作为标识符
    'dot-notation': 2,                                 // 要求尽可能使用点号, obj.xxx
    'padding-line-between-statements': [2,
      { blankLine: 'always', prev: 'block-like', next: '*' }, // 块语句之前需要空行
      { blankLine: 'always', prev: '*', next: 'block-like' }, // 块语句之后需要空行
      { blankLine: 'always', prev: 'export', next: '*' },     // export 语句之前需要空行
      { blankLine: 'always', prev: '*', next: 'export' },     // export 语句之后需要空行
    ],

    // 3. import
    'import/prefer-default-export': 2,  // 在只有一个导出的模块里, 用 export default
    'import/first': 2,                  // 10.7 import 放在其他所有语句之前
  },
};
